### PR TITLE
Fix lazy string format

### PR DESCRIPTION
**Pull request by rokonec**
In couple of places we have not leveraged LazyFormattedBuildEventArgs properly.
With these changes strings would not be necessary to format, in cases when logging is limited like `/clp:PerformanceSummary`

Related issue #2700

### CODE REVIEW

**Comment by rokonec** [comment]
After discussion with Rainer we have decided to delete `#ifdef VALIDATERESOURCESTRINGS` sections as it has not been used since 2006

**Review by BenVillalobos — COMMENTED** [review]
Just making sure I understand this optimization at a high level.
1. `TaskLoggingHelper` uses `BuildMessageEventArgs` to log messages, and `BuildMessageEventArgs` is a `LazyFormattedBuildEventArgs`.
2. `GetResourceString` no longer auto-formats because `TaskLoggingHelper` ultimately uses a `LazyFormattedBuildEventArgs`
3. Any calls to `TaskLoggingHelper.LogMessage` that used to pass in a formatted string have been modified to pass in the resource strings unformatted, thus preventing an unnecessary format

If my understanding is correct, LGTM. I double checked and there shouldn't be any other calls to `LogMessage` that pass in a formatted string after this PR merged.

**Comment by rokonec** [comment]
@BenVillalobos correct. The main idea here is to leverage LazyFormattedBuildEventArgs to safe string Format and its costly string allocation in cases when considerable subset of logs are disabled.
I have done PerfView profiling and it proved it worked.

**Review by BenVillalobos — APPROVED** [review]

File: src/Framework/LazyFormattedBuildEventArgs.cs [hunk 1]
```diff
@@ -199,45 +199,6 @@ private static string FormatString(CultureInfo culture, string unformatted, para
             if ((args?.Length > 0))
             {
 #if DEBUG
-
-#if VALIDATERESOURCESTRINGS
```
Comment (rainersigwald) [inline src/Framework/LazyFormattedBuildEventArgs.cs:203]: Nit: I'd ideally like cleanup like this in a separate commit so we can easily separate it out from the more interesting work in the PR.

**Review by rainersigwald — APPROVED** [review]
Looks great, thanks!

**Review by ladipro — CHANGES_REQUESTED** [review]
`TaskLoggingHelper` is part of public API surface and - strictly technically - the way it's refactored in this PR is not 100% safe. Say that someone has a derived class:

```C#
class MyLoggingHelper : TaskLoggingHelper
{
    public override string GetResourceMessage(string resourceName)
    {
        return "Hello: " + FormatResourceString(resourceName, null);
    }
}
```

Now with the change in this PR calling `GetResourceMessage` on such a class will crash with SO exception while previously it worked perfectly fine.

Am I being paranoid? Maybe.

Here's a less contrived example: Changing the formatting behavior can currently be achieved by overriding only `FormatResourceString` because then you get `GetResourceMessage` for free. With your change it is no longer true. Actually, case in point, check out `TaskLoggingHelperExtension`. This class would be broken exactly like so because `GetResourceMessage` wouldn't get the overriden behavior.

File: src/Framework/LazyFormattedBuildEventArgs.cs (hunk 1 above)
Comment (rokonec) [reply src/Framework/LazyFormattedBuildEventArgs.cs:203]: Understand

File: src/Shared/TaskLoggingHelper.cs [hunk 2]
```diff
@@ -272,6 +271,13 @@ public void LogMessage(MessageImportance importance, string message, params obje
         {
             // No lock needed, as BuildEngine methods from v4.5 onwards are thread safe.
             ErrorUtilities.VerifyThrowArgumentNull(message, nameof(message));
+#if DEBUG
+            if (messageArgs?.Length > 0)
+            {
+                // Verify that message can be formated using given arguments
```
[the remark this answers is not in the document]
Comment (rokonec) [reply src/Shared/TaskLoggingHelper.cs:277]: Fixed

**Comment by rokonec** [comment]
@ladipro This is very good catch. I will fix it.

But truth to be said, the current implementation is counter intuitive -  at least to me - as one would expect `FormatResourceSttring` calls `GetResourceMessage` and not vice versa, but since this 'silence' contract has been published, there is not much we can do about it as there could indeed be users relying on it - and as you pointed we do we in `TaskLoggingHelperExtension`.

**Review by ladipro — APPROVED** [review]
Thank you!

**Pull request merged**